File: Travelling Salesman Problem 3_python/src/Maze.py
import sys
import traceback

from Direction import Direction
from SurroundingPheromone import SurroundingPheromone


# Class that holds all the maze data. This means the pheromones, the open and blocked tiles in the system as
# well as the starting and end coordinates.


class Maze:

    # Constructor of a maze
    # @param walls int array of tiles accessible (1) and non-accessible (0)
    # @param width width of Maze (horizontal)
    # @param length length of Maze (vertical)
    def __init__(self, walls, width, length):
        self.pheromones = []
        self.walls = walls
        self.length = length
        self.width = width
        self.start = None
        self.end = None
        self.initialize_pheromones()

    def initialize_pheromones(self):
        """Initialize pheromones to a start value. """
        for i in range(self.width):
            self.pheromones.append([])
            for _ in range(self.length):
                self.pheromones[i].append(0)

    # Reset the maze for a new shortest path problem.
    def reset(self):
        self.initialize_pheromones()

    def get_possible_actions(self, position):
        options = []
        for i in [Direction.north, Direction.east, Direction.south, Direction.west]:
            possible_location = position.add_direction(i)
            if self.in_bounds(possible_location) \
                    and self.walls[possible_location.get_x()][possible_location.get_y()] == 1:
                options.append(i)
        return options

    # @staticmethod
    # def add_dir(x,y, dir):
    #     return
    # Update the pheromones along a certain route according to a certain Q
    # @param r The route of the ants
    # @param Q Normalization factor for amount of dropped pheromone
    def add_pheromone_route(self, route, q):
        if len(route.route) == 0:
            return
        positions = []
        ph_to_add = (1 / route.size()) * q
        position = route.get_start()
        directions = route.get_route()
        self.pheromones[position.get_x()][position.get_y()] += ph_to_add
        for direction in directions:
            position = position.add_direction(direction)
            # if position not in positions:
            self.pheromones[position.get_x()][position.get_y()] += ph_to_add
                # positions.append(position)
        return

    # Update pheromones for a list of routes
    # @param routes A list of routes
    # @param Q Normalization factor for amount of dropped pheromone
    def add_pheromone_routes(self, routes, q):
        for r in routes:
            self.add_pheromone_route(r, q)

    # Evaporate pheromone
    # @param rho evaporation factor
    def evaporate(self, rho, factor):
        max = 0
        for i in range(self.width):
            for j in range(self.length):
                self.pheromones[i][j] *= rho
                if self.pheromones[i][j] > max:
                    max = self.pheromones[i][j]
        minconstant = factor * max
        for i in range(self.width):
            for j in range(self.length):
                if self.pheromones[i][j] <= minconstant:
                    self.pheromones[i][j] = 0
        # Add statement for minimum amout of pheromone
        return

    # Width getter
    # @return width of the maze
    def get_width(self):
        return self.width

    # Length getter
    # @return length of the maze
    def get_length(self):
        return self.length

    def get_surrounding_pheromone(self, position):
        """
        Returns a the amount of pheromones on the neighbouring positions (N/S/E/W).
        :type position: Coordinate.Coordinate
        :param position: The position to check the neighbours of.
        :return: the pheromones of the neighbouring positions.
        """
        east = self.get_pheromone(position.add_direction(Direction.east))
        north = self.get_pheromone(position.add_direction(Direction.north))
        west = self.get_pheromone(position.add_direction(Direction.west))
        south = self.get_pheromone(position.add_direction(Direction.south))
        return SurroundingPheromone(north, east, south, west)

    # Pheromone getter for a specific position. If the position is not in bounds returns 0
    # @param pos Position coordinate
    # @return pheromone at point
    def get_pheromone(self, pos):
        if not self.in_bounds(pos):
            return 0
        return self.pheromones[pos.get_x()][pos.get_y()]

    # Check whether a coordinate lies in the current maze.
    # @param position The position to be checked
    # @return Whether the position is in the current maze
    def in_bounds(self, position):
        return position.x_between(0, self.width) and position.y_between(0, self.length)

    # Representation of Maze as defined by the input file format.
    # @return String representation
    def __str__(self):
        string = ""
        string += str(self.width)
        string += " "
        string += str(self.length)
        string += " \n"
        for y in range(self.length):
            for x in range(self.width):
                string += str(self.walls[x][y])
                string += " "
            string += "\n"
        return string

    # Method that builds a mze from a file
    # @param filePath Path to the file
    # @return A maze object with pheromones initialized to 0's inaccessible and 1's accessible.
    @staticmethod
    def create_maze(file_path):
        try:
            f = open(file_path, "r")
            lines = f.read().splitlines()
            dimensions = lines[0].split(" ")
            width = int(dimensions[0])
            length = int(dimensions[1])

            # make the maze_layout
            maze_layout = []
            for x in range(width):
                maze_layout.append([])

            for y in range(length):
                line = lines[y + 1].split(" ")
                for x in range(width):
                    if line[x] != "":
                        state = int(line[x])
                        maze_layout[x].append(state)
            print("Ready reading maze file " + file_path)
            return Maze(maze_layout, width, length)
        except FileNotFoundError:
            print("Error reading maze file " + file_path)
            traceback.print_exc()
            sys.exit()

File: Travelling Salesman Problem 3_java/src/main/java/Maze.java
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Scanner;

/**
 * Class that holds all the maze data. This means the pheromones, the open and blocked tiles in the system as
 * well as the starting and end coordinates.
 */
public class Maze {
    private int width;
    private int length;
    private int[][] walls;
    private double[][] pheromones;
//    private Coordinate start;
//    private Coordinate end;

    /**
     * Constructor of a maze
     *
     * @param walls  int array of tiles accessible (1) and non-accessible (0)
     * @param width  width of Maze (horizontal)
     * @param length length of Maze (vertical)
     */
    public Maze(int[][] walls, int width, int length) {
        this.walls = walls;
        this.length = length;
        this.width = width;
        initializePheromones();
    }

    /**
     * Method that builds a mze from a file
     *
     * @param filePath Path to the file
     * @return A maze object with pheromones initialized to 0's inaccessible and 1's accessible.
     */
    public static Maze createMaze(String filePath) throws FileNotFoundException {
        Scanner scan = new Scanner(new FileReader(filePath));
        int width = scan.nextInt();
        int length = scan.nextInt();
        int[][] mazeLayout = new int[width][length];
        for (int y = 0; y < length; y++) {
            for (int x = 0; x < width; x++) {
                mazeLayout[x][y] = scan.nextInt();
            }
        }
        scan.close();
        return new Maze(mazeLayout, width, length);
    }

    /**
     * Initialize pheromones to a start value.
     */
    private void initializePheromones() {
        this.pheromones = new double[this.width][this.length];
        for (int i = 0; i < this.width; i++) {
            for (int j = 0; j < this.length; j++) {
                this.pheromones[i][j] = 10;
            }
        }
    }

    /**
     * Reset the maze for a new shortest path problem.
     */
    public void reset() {
        initializePheromones();
    }

    /**
     * Update the pheromones along a certain route according to a certain Q
     *
     * @param r The route of the ants
     * @param Q Normalization factor for amount of dropped pheromone
     */
    public void addPheromoneRoute(Route r, double Q) {
        double phToAdd = (1.0 / r.size()) * Q;
//        System.out.println(phToAdd);
        Coordinate pos = r.getStart();
        this.pheromones[pos.getX()][pos.getY()] += phToAdd;
        HashSet<Coordinate> poss = new HashSet<>(r.size());
        for (Direction d : r.getRoute()) {
            pos = pos.add(d);
            if (poss.add(pos)) {
                this.pheromones[pos.getX()][pos.getY()] += phToAdd;
            }
        }

    }

    /**
     * Update pheromones for a list of routes
     *
     * @param routes A list of routes
     * @param Q      Normalization factor for amount of dropped pheromone
     */
    public void addPheromoneRoutes(List<Route> routes, double Q) {
        for (Route r : routes) {
            addPheromoneRoute(r, Q);
        }
    }

    /**
     * Evaporate pheromone
     *
     * @param rho    evaporation factor
     * @param factor
     */
    public void evaporate(double rho, double factor) {
        double max = 0;
        for (int i = 0; i < this.width; i++) {
            for (int j = 0; j < this.length; j++) {
                this.pheromones[i][j] *= (1.0 - rho);
//                if (this.pheromones[i][j] > max) {
//                    max = this.pheromones[i][j];
//                }
            }
        }
//        double minconstant = max * factor;
//        for (int i = 0; i < this.width; i++) {
//            for (int j = 0; j < this.length; j++) {
//                if (this.pheromones[i][j] < minconstant)
//                    this.pheromones[i][j] = 0;
//            }
//        }
    }

    /**
     * Width getter
     *
     * @return width of the maze
     */
    public int getWidth() {
        return width;
    }

    /**
     * Length getter
     *
     * @return length of the maze
     */
    public int getLength() {
        return length;
    }

    /**
     * Returns a the amount of pheromones on the neighbouring positions (N/S/E/W).
     *
     * @param position The position to check the neighbours of.
     * @return the pheromones of the neighbouring positions.
     */
    public SurroundingPheromone getSurroundingPheromone(Coordinate position) {
        double east = this.getPheromone(position.add(Direction.East));
        double north = this.getPheromone(position.add(Direction.North));
        double south = this.getPheromone(position.add(Direction.South));
        double west = this.getPheromone(position.add(Direction.West));
        return new SurroundingPheromone(north, east, south, west);
    }

    public List<Direction> getValidActions(Coordinate pos) {
        List<Direction> pactions = new ArrayList<>(4);
        for (int i = 0; i < 4; i++) {
            Coordinate newPos = pos.add(Direction.intToDir(i));
            if (this.isWalkAble(newPos)) {
                pactions.add(Direction.intToDir(i));
            }
        }
        return pactions;

    }

    private boolean isWalkAble(Coordinate p) {
        if (!this.inBounds(p)) {
            return false;
        }
        return walls[p.getX()][p.getY()] == 1;
    }

    /**
     * Pheromone getter for a specific position. If the position is not in bounds returns 0
     *
     * @param pos Position coordinate
     * @return pheromone at point
     */
    private double getPheromone(Coordinate pos) {
        if (!this.inBounds(pos)) {
            return 0;
        }
        return this.pheromones[pos.getX()][pos.getY()];
    }

    /**
     * Check whether a coordinate lies in the current maze.
     *
     * @param position The position to be checked
     * @return Whether the position is in the current maze
     */
    public boolean inBounds(Coordinate position) {
        return position.xBetween(0, width) && position.yBetween(0, length);
    }

    /**
     * Representation of Maze as defined by the input file format.
     *
     * @return String representation
     */
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(width);
        sb.append(' ');
        sb.append(length);
        sb.append(" \n");
        for (int y = 0; y < length; y++) {
            for (int x = 0; x < width; x++) {
                sb.append(walls[x][y]);
                sb.append(' ');
            }
            sb.append("\n");
        }
        return sb.toString();
    }

    public void printPheromones(Route r, PathSpecification spec) {
        StringBuilder sb = new StringBuilder("Pheromones: \n");
        for (int i = 0; i < this.length; i++) {
            sb.append("|");
            for (int j = 0; j < this.width; j++) {
                sb.append(String.format("%-8.0f ", pheromones[j][i]));
            }
            sb.append("|\n");
        }
        try {
            if (!Files.exists(Paths.get("output"))) {
                Files.createDirectory(Paths.get("output"));
            }
            Files.write(Paths.get(String.format("output/pheromones_%s-%s(%d).txt",
                spec.getStart(), spec.getEnd(), r.size())), sb.toString().getBytes());
        } catch (IOException e) {
            e.printStackTrace();
        }
//        System.out.println(sb.toString());
    }
}
